### PR TITLE
[FEATURE] Adds support for attaching MPM particles to rigid links.

### PR DESCRIPTION
**Pull request by YilingQiao**
Adds support for attaching MPM particles to rigid links using soft constraints, similar to the existing FEM vertex constraint system.

https://github.com/user-attachments/assets/37e6dc5e-5630-44b4-8645-314c388acff3

## Changes

**`genesis/options/solvers.py`**
- Add `enable_particle_constraints` option to `MPMOptions` (default: False)

**`genesis/engine/solvers/mpm_solver.py`**
- Add `particle_constraint_info` struct for constraint data
- Add `init_constraints()` for lazy field initialization
- Add `_kernel_set_particle_constraints()` and `_kernel_remove_particle_constraints()` kernels
- Add `apply_particle_constraints()` kernel with `ti.static` branching for zero overhead when disabled
- Call constraint kernel in `substep_post_coupling()` after G2P

**`genesis/engine/entities/mpm_entity.py`**
- Add `get_particles_in_bbox()` helper to find particles in a region
- Add `set_particle_constraints()` to attach particles to a rigid link
- Add `remove_particle_constraints()` to detach particles

## Usage

```python
scene = gs.Scene(
    mpm_options=gs.options.MPMOptions(
        enable_particle_constraints=True,
        ...
    )
)

# Find particles in attachment region
particles = mpm_entity.get_particles_in_bbox(bbox_min, bbox_max)

# Attach to rigid link with spring stiffness
mpm_entity.set_particle_constraints(particles, link=robot.links[0], stiffness=1e5)
```

## Tests
- `tests/test_mpm.py`: Unit tests for constraint API and behavior

## Example
- `examples/coupling/rigid_mpm_attachment.py`: Demo of MPM cube attached to oscillating rigid box

## Checklist:

- [ ] I read the **CONTRIBUTING** document.
- [ ] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [ ] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [ ] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [ ] I tested my changes and added instructions on how to test it for reviewers.

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: tests/test_deformable_physics.py [hunk 1]
```diff
@@ -221,6 +221,71 @@ def test_deformable_parallel(show_viewer):
...
+    rigid_box = scene.add_entity(gs.morphs.Box(pos=(0, 0, 0.5), size=(0.1, 0.1, 0.05), fixed=True))
+    mpm_cube = scene.add_entity(
+        material=gs.materials.MPM.Elastic(E=1e4, nu=0.3, rho=1000),
+        morph=gs.morphs.Box(pos=(0, 0, 0.3), size=(0.1, 0.1, 0.1)),
+    )
+    scene.build()
+
+    particles = mpm_cube.get_particles_in_bbox([-0.05, -0.05, 0.35], [0.05, 0.05, 0.4])
+    with pytest.raises(gs.GenesisException, match="enable_particle_constraints"):
+        mpm_cube.set_particle_constraints(particles, link=rigid_box.links[0], stiffness=1e4)
```
Comment (YilingQiao) [inline tests/test_deformable_physics.py]: enable_particle_constraints=False by default since the data structure for constraints is hugh

File: genesis/engine/entities/mpm_entity.py [hunk 2]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+            (poss[:, 0] >= bbox_min[0])
+            & (poss[:, 0] <= bbox_max[0])
+            & (poss[:, 1] >= bbox_min[1])
+            & (poss[:, 1] <= bbox_max[1])
+            & (poss[:, 2] >= bbox_min[2])
+            & (poss[:, 2] <= bbox_max[2])
+        )
+
+        particles_idx_local = torch.nonzero(in_bbox, as_tuple=False).squeeze(-1)
+        return particles_idx_local
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: `torch.nonzero` should be avoided because it forces GPU sync. Try supporting boolean mask all the way for efficiency.

File: genesis/engine/entities/mpm_entity.py [hunk 3]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+
+        self._solver._kernel_set_particle_constraints(
+            self._sim.cur_substep_local,
+            particles_idx,
+            envs_idx,
+            stiffness,
+            link.idx,
+            link_pos,
+            link_quat,
+        )
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py:659]: Add support of zero copy with boolean masking.

File: genesis/engine/entities/mpm_entity.py [hunk 4]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+        if particles_idx_local is None:
+            self._solver.particle_constraints.is_constrained.fill(False)
+            self._solver.particle_constraints.link_idx.fill(-1)
+            return
+
+        envs_idx = self._scene._sanitize_envs_idx(envs_idx)
+        particles_idx_local = self._sanitize_particles_idx_local(particles_idx_local, envs_idx)
+        particles_idx = particles_idx_local + self._particle_start
+
+        self._solver._kernel_remove_particle_constraints(particles_idx, envs_idx)
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: Add support of zero-copy with boolean masking.

File: genesis/options/solvers.py [hunk 5]
```diff
@@ -431,6 +431,8 @@ class MPMOptions(Options):
         Lower bound of the simulation domain. Defaults to (-1.0, -1.0, 0.0).
     upper_bound : tuple, shape (3,), optional
         Upper bound of the simulation domain. Defaults to (1.0, 1.0, 1.0).
+    enable_particle_constraints : bool, optional
+        Whether to enable particle constraints for attaching MPM particles to rigid links. Defaults to False.
```
Comment (duburcqa) [inline genesis/options/solvers.py]: I'm not convinced it is relevant to have a flag to disable this feature. What is supposed to be the drawback of enabling this systematically? Flags complexifying the control flow should only be added if necessary because it makes the codebase harder to maintain and comprehend.

Comment (duburcqa) [reply genesis/options/solvers.py]: If you want to maintain efficiency at all cost, I would rather manage some external flag `self._has_particles_constraint` that would be set/unset by `set_particles_constraint` and `remove_particles_constraint`. But I would not recommend this unless the performance drop is larger than 8% or something. 

edit: you could just use `_constraints_initialized`, it should be good enough.

File: genesis/engine/entities/mpm_entity.py [hunk 6]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+        Parameters
+        ----------
+        particles_idx_local : array_like, optional
+            Local particle indices to unconstrain. If None, removes all constraints.
+        envs_idx : array_like, optional
+            Environment indices. If None, applies to all environments.
+        """
+        if not self._solver._constraints_initialized:
+            gs.logger.warning("Ignoring remove_particle_constraints; constraints have not been initialized.")
+            return
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py:672]: I don't think this warning is relevant.

File: genesis/engine/entities/mpm_entity.py [hunk 7]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+
+        # Find particles in bounding box
+        in_bbox = (
+            (poss[:, 0] >= bbox_min[0])
+            & (poss[:, 0] <= bbox_max[0])
+            & (poss[:, 1] >= bbox_min[1])
+            & (poss[:, 1] <= bbox_max[1])
+            & (poss[:, 2] >= bbox_min[2])
+            & (poss[:, 2] <= bbox_max[2])
+        )
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: Please vectorise these computations:
```python
in_bbox = (bbox_min <= poss & poss <= bbox_max).all(dim=-1)
```

File: genesis/engine/entities/mpm_entity.py [hunk 8]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+        envs_idx : array_like, optional
+            Environment indices. If None, uses first environment for position query.
+
+        Returns
+        -------
+        particles_idx_local : torch.Tensor, shape (n_in_bbox,)
+            Local particle indices within the bounding box.
+        """
+        bbox_min = torch.tensor(bbox_min, dtype=gs.tc_float, device=gs.device)
+        bbox_max = torch.tensor(bbox_max, dtype=gs.tc_float, device=gs.device)
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: Use `torch.as_tensor`

File: genesis/engine/entities/mpm_entity.py [hunk 9]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+
+        # Get particle positions (use first env if batched)
+        poss = self.get_particles_pos(envs_idx=envs_idx)
+        if poss.ndim == 3:
+            poss = poss[0]  # Use first environment
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: This is not acceptable. Please add support of batching. This is straightforward if you return a boolean mask as I requested anyway.

File: genesis/engine/entities/mpm_entity.py [hunk 10]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+        """
+        Attach MPM particles to a rigid link using soft constraints.
+
+        The particles will be pulled toward their relative position on the link
+        using spring forces with critical damping.
+
+        Parameters
+        ----------
+        particles_idx_local : array_like
+            Local particle indices to constrain.
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: Replace this by boolean mask.

File: genesis/engine/entities/mpm_entity.py [hunk 11]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+        if not self._solver._enable_particle_constraints:
+            gs.raise_exception(
+                "Particle constraints are disabled. Set 'enable_particle_constraints=True' in MPMOptions."
+            )
+
+        if not self._solver._constraints_initialized:
+            self._solver.init_constraints()
+
+        if not isinstance(link, RigidLink):
+            gs.raise_exception("Only RigidLink is supported for particle constraints.")
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: Move this check on top.

File: genesis/engine/entities/mpm_entity.py [hunk 12]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+            & (poss[:, 1] <= bbox_max[1])
+            & (poss[:, 2] >= bbox_min[2])
+            & (poss[:, 2] <= bbox_max[2])
+        )
+
+        particles_idx_local = torch.nonzero(in_bbox, as_tuple=False).squeeze(-1)
+        return particles_idx_local
+
+    @gs.assert_built
+    def set_particle_constraints(self, particles_idx_local, link, stiffness=1e4, envs_idx=None):
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: Do not pass link but rather link_idx. We do not want to pass complex objects as input arguments. It is just asking for trouble in terms of memory management, and it is harder to keep track of the simulation state.

File: genesis/engine/entities/mpm_entity.py [hunk 13]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+        using spring forces with critical damping.
+
+        Parameters
+        ----------
+        particles_idx_local : array_like
+            Local particle indices to constrain.
+        link : RigidLink
+            The rigid link to attach particles to.
+        stiffness : float, optional
+            Spring stiffness for the constraint. Defaults to 1e4.
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: I'm not in favour of adding default value. It is very much system specific, there is no good universal default.

File: genesis/engine/entities/mpm_entity.py [hunk 14]
```diff
@@ -581,3 +581,131 @@ def get_free(self):
...
+            Environment indices. If None, applies to all environments.
+        """
+        if not self._solver._constraints_initialized:
+            gs.logger.warning("Ignoring remove_particle_constraints; constraints have not been initialized.")
+            return
+
+        if particles_idx_local is None:
+            self._solver.particle_constraints.is_constrained.fill(False)
+            self._solver.particle_constraints.link_idx.fill(-1)
+            return
```
Comment (duburcqa) [inline genesis/engine/entities/mpm_entity.py]: Use zero-copy for this.

File: genesis/engine/solvers/mpm_solver.py [hunk 15]
```diff
@@ -163,6 +165,25 @@ def init_vvert_fields(self):
...
+            is_constrained=gs.ti_bool,  # whether particle is constrained
+            target_pos=gs.ti_vec3,  # target position for the constraint
+            stiffness=gs.ti_float,  # spring stiffness
+            link_idx=gs.ti_int,  # index of the rigid link (-1 if not linked)
+            link_local_pos=gs.ti_vec3,  # offset from link origin in link's local frame
+        )
+
+        self.particle_constraints = particle_constraint_info.field(
+            shape=(self._n_particles, self._B), needs_grad=False, layout=ti.Layout.AOS
+        )
```
Comment (duburcqa) [inline genesis/engine/solvers/mpm_solver.py:188]: Add a check to make sure the index fits in memory -> self._n_particles * self._B * 3 < max(int32)

File: tests/test_deformable_physics.py [hunk 16]
```diff
@@ -221,6 +221,71 @@ def test_deformable_parallel(show_viewer):
     assert_allclose(water.get_particles_vel(), 0.0, atol=5e-2)
 
 
+@pytest.mark.required
+def test_mpm_particle_constraints_disabled(show_viewer):
+    """Test that particle constraints raise error when disabled."""
+    scene = gs.Scene(
+        sim_options=gs.options.SimOptions(dt=2e-3, substeps=10),
```
Comment (duburcqa) [inline tests/test_deformable_physics.py]: One line per option in unit tests.

File: tests/test_deformable_physics.py [hunk 17]
```diff
@@ -221,6 +221,71 @@ def test_deformable_parallel(show_viewer):
...
+        mpm_options=gs.options.MPMOptions(
+            lower_bound=(-0.5, -0.5, 0.0),
+            upper_bound=(0.5, 0.5, 1.0),
+            grid_density=64,
+        ),
+        show_viewer=show_viewer,
```
Comment (duburcqa) [inline tests/test_deformable_physics.py]: Disable FPS counter.

File: tests/test_deformable_physics.py [hunk 18]
```diff
@@ -221,6 +221,71 @@ def test_deformable_parallel(show_viewer):
...
+    )
+    scene.add_entity(gs.morphs.Plane())
+    rigid_box = scene.add_entity(gs.morphs.Box(pos=(0, 0, 0.5), size=(0.1, 0.1, 0.05), fixed=True))
+    mpm_cube = scene.add_entity(
+        material=gs.materials.MPM.Elastic(E=1e4, nu=0.3, rho=1000),
+        morph=gs.morphs.Box(pos=(0, 0, 0.3), size=(0.1, 0.1, 0.1)),
+    )
+    scene.build()
```
Comment (duburcqa) [inline tests/test_deformable_physics.py]: One line per option in unit tests

Comment (duburcqa) [inline tests/test_deformable_physics.py]: Make sure that batching is supported.

File: tests/test_deformable_physics.py (hunk 1 above)
Comment (duburcqa) [reply tests/test_deformable_physics.py]: > enable_particle_constraints=False by default since the data structure for constraints is hugh

I don't think it is a good argument since constraint data fields are lazily initialised only if necessary (which is great).

File: tests/test_deformable_physics.py [hunk 19]
```diff
@@ -221,6 +221,71 @@ def test_deformable_parallel(show_viewer):
...
+
+    # Attach and test following
+    mpm_cube.set_particle_constraints(attached, link=rigid_box.links[0], stiffness=1e5)
+    initial_rigid_pos = rigid_box.get_pos().clone()
+    initial_mpm_x = mpm_cube.get_particles_pos()[attached, 0].mean()
+
+    pos_diff = torch.tensor([0.2, 0, 0], device=gs.device)
+    rigid_box.set_pos(initial_rigid_pos + pos_diff, zero_velocity=False)
+    for _ in range(50):
+        scene.step()
```
Comment (duburcqa) [inline tests/test_deformable_physics.py]: 50 steps is too much. Can you reduce this? Otherwise I'm afraid this test will be too slow.

File: tests/test_deformable_physics.py [hunk 20]
```diff
@@ -221,6 +221,71 @@ def test_deformable_parallel(show_viewer):
...
+    )
+    scene.build()
+
+    particles = mpm_cube.get_particles_in_bbox([-0.05, -0.05, 0.35], [0.05, 0.05, 0.4])
+    with pytest.raises(gs.GenesisException, match="enable_particle_constraints"):
+        mpm_cube.set_particle_constraints(particles, link=rigid_box.links[0], stiffness=1e4)
+
+
+@pytest.mark.required
+def test_mpm_particle_constraints(show_viewer):
```
Comment (duburcqa) [inline tests/test_deformable_physics.py:225]: You should flag this test as slow if it is (>200s). Beware that it means that it will only run on production CI, so only do this if necessary.

Comment (YilingQiao) [reply tests/test_deformable_physics.py:225]: it's not slow on my machine, but I'm happy to add the tag to make our generic CI not too slow

File: genesis/engine/solvers/mpm_solver.py (hunk 15 above)
Comment (YilingQiao) [reply genesis/engine/solvers/mpm_solver.py:188]: done

**Comment by YilingQiao** [comment]
Comments are addressed

File: tests/test_deformable_physics.py (hunk 20 above)
Comment (duburcqa) [reply tests/test_deformable_physics.py:225]: > it's not slow on my machine, but I'm happy to add the tag to make our generic CI not too slow

The speed must be benched on the generic CI runners. Not your machine. So you need to run the entire CI pipeline without "slow" mark at least once, otherwise you will never know since it will be skipped.

**Comment by duburcqa** [comment]
Note that you are still not using zero-copy but I will let that slip in I don't think it is a bottleneck.
